### DISCUSSION TITLE
Logging to file and silencing liesel messages

### DISCUSSION THREAD

**Discussion by jeli2103** [liesel-devs/liesel#163] [Q&A] [answered]
Hey Liesels,
I would like to log all liesel messages (plus own messages) to a file, but want to avoid that messages are printed to the console. I have tried the following:

```python 
"""python script"""
import logging
import liesel.model as lsl
import liesel.goose as gs

# loggers
logger_liesel = logging.getLogger('liesel')
logger_liesel.setLevel(logging.INFO)
logger_gr = logging.getLogger('groupie')
logger_gr.setLevel(logging.INFO)

# create file handler which logs to file
fh = logging.FileHandler('liesel_groupie.log')
fh.setLevel(logging.INFO)

# create formatter and add
formatter = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
fh.setFormatter(formatter)

# add handler
logger_liesel.addHandler(fh)
logger_gr.addHandler(fh)

logger_gr.info('nothing is impossible')

# start liesel goose engine: engine.sample_all_epochs()  

```

In my case the file liesel_groupie.log looks good. The console, however, does not have 'nothing is impossible', but all liesel messages. How can I silence them as well?

**Comment by jobrachem** [comment, marked as answer]
You can basically remove the log handler that prints messages to the console. You can use a function like this:

```python
import logging

def remove_first_liesel_log_handler():
    liesel_logger = logging.getLogger("liesel")
    liesel_stdout = liesel_logger.handlers[0]
    liesel_logger.removeHandler(liesel_stdout)
```

Then, to remove logging to the console and instead add a log handler that logs to a file, you can do the following:

```python
from liesel.logging import add_file_handler

add_file_handler(path="your/logfile.log", level="info")
remove_first_liesel_log_handler()
```

**Reply by jeli2103** [reply]
Great, so easy to remove then. Thank you. 
If working only on liesel level messages, this is the solution. In my case I want to mix own messages into the same log file. In that case you need to add the same handler to the different loggers (I guess). So it would be:

```python

"""python script"""
import logging
import liesel.model as lsl
import liesel.goose as gs

def remove_first_liesel_log_handler():
    liesel_logger = logging.getLogger("liesel")
    liesel_stdout = liesel_logger.handlers[0]
    liesel_logger.removeHandler(liesel_stdout)

# loggers
logger_liesel = logging.getLogger('liesel')
logger_liesel.setLevel(logging.INFO)
logger_gr = logging.getLogger('groupie')
logger_gr.setLevel(logging.INFO)

# create file handler which logs to file
fh = logging.FileHandler('liesel_groupie.log')
fh.setLevel(logging.INFO)

# create formatter and add
formatter = logging.Formatter('%(name)s - %(levelname)s - %(message)s')
fh.setFormatter(formatter)

# add handler
logger_liesel.addHandler(fh)
logger_gr.addHandler(fh)

# rm messages from stdout (only for liesel logger required)
remove_first_liesel_log_handler()

logger_gr.info('nothing is impossible')
logger_liesel.info('nothing is impossible=liesel')

```

So now the log file should contain 

```
groupie - INFO - nothing is impossible
liesel - INFO - nothing is impossible=liesel
```

**Reply by jeli2103** [reply]
If the function "add_file_handler" would have an option like return=True and I could work with the returned object, would be useful in that case...

**Reply by jobrachem** [reply]
I think you can also add a new file handler that points to the same logfile to your own logger. You can even use the `add_file_handler` function for that:

```python
from liesel.logging import add_file_handler

add_file_handler(path="your/logfile.log", level="info")
add_file_handler(path="your/logfile.log", level="info", logger="groupie")
remove_first_liesel_log_handler()
```

**Reply by jobrachem** [reply]
P.S.: Even though, we currently do not include the documentation for `add_file_handler`, the function's source code is indeed documented. You can have a look here: [add_file_handler](https://github.com/liesel-devs/liesel/blob/727bd853407998c0124911e88e87c9b652c79e9b/liesel/logging.py#L74)

**Reply by jeli2103** [reply]
oh. yes. Actually it works. I don't know why it didn't before. It is indeed a helpful function. Thank you. 
One needs to be aware of giving the absolute path to the path argument of add_file_handler, path='/home/...'. There seems to be a problem, however, if you set a path starting with a slash and it is not your absolute path. Then no error occurs and no files are saved (I unintentionally did so and was confused about the whereabouts of the files).

**Reply by jobrachem** [reply]
@liesel-groupie I opened an issue addressing your report on the behavior of the add_file_handler function: #164

**Reply by jobrachem** [reply]
Coming back: Starting a path with a slash marks it as absolute. See #164.
